D-Freezer/freezer.md

```diff
@@ -0,0 +1,186 @@
+
+By convention, Linux scheduler-specific wrapper structures `struct <sched_class>_rq`. For example, the CFS class defines a `struct cfs_rq` which is then declared inside of `struct rq` as `struct cfs_rq cfs`.
+
+snippet is taken from `linux/kernel/sched/sched.h`
+```
+struct cfs_rq {
+	struct load_weight	load;
+	unsigned long		runnable_weight;
+	unsigned int		nr_running;
+	unsigned int		h_nr_running;
+       /* code omitted */
+};
+
+struct rq {
+        /* code omitted */
+	struct cfs_rq		cfs;
+	struct rt_rq		rt;
+	struct dl_rq		dl;
+        /* code omitted */
+}
+```
+## The `freezer_rq`
+
+At this point, you have probably guessed that you will need to do the same thing for the freezer. You are right. The `freezer_rq` should include the head of the freezer runqueue. Additionally, you may need to include some bookkeeping variables. Word of advice, think of what you would actually need and don't have anything extra (it should be pretty simple). 
+
+## The `sched_freezer_entity`
+
+Now that you have the `struct rq` setup, you need to have some mechanism to join your `task_struct`s into the queue. Here, too, you can't just include a `list_head node` to add a task onto the scheduler-specific runqueue because you'll need additional bookkeeping. As you have probably guessed, we are going to wrap the list_head and all the bookkeeping variables in its own struct. In Linux, we name these structs `sched_{class}_entity` (one exception is that CFS names this `sched_entity`). For example, the Real-time scheduling class calls it `sched_rt_entity`. We will name ours `struct sched_freezer_entity`. Again, make sure you only include what you need in this struct. 
+
+With all this setup, here is what the final picture would look like:
+
+<div align='center'>
+    <img src='./freezer.png'/><br/>
+</div>
+
+In the picture above, the two structs on the far left represent a system with two CPUs. I colored these blue and green for two reasons: 1) to distinguish them from each other, and 2) to show different `task_structs` linked on one `siblings` linked-list can run on separate CPUs. 
+
+## The `sched_class`
+
+At this point, we have set up the data structures, but we are still not done. We now need to implement the freezer functionality to let the Kernel to use the freezer as a scheduler. 
+Here is the problem. Say we have a CFS task about to return from main(), we need to call the CFS `decueue_task()` to remove it from the CFS queue. How can we ensure that the Kernel will call the CFS implementation of `dequeue_task()`? The answer is `struct sched_class` defined in `linux/kernel/sched/sched.h`. Here is what the structure looks like.
```

Comment:
dequeue_task

change "we need to" to "the OS needs to", otherwise, they may think they have to implement this

"how can we ensure that the Kernel" -> "how can we ensure that the OS"